File: 1297-maximum-number-of-balloons/maximum-number-of-balloons.cpp
class Solution {
public:
    int maxNumberOfBalloons(string text) 
    {
        int b = 0, a = 0, l = 0, o = 0, n = 0;

        for(char ch : text)
        {
            switch(ch)
            {
                case 'b': b++; break;
                case 'a': a++; break;
                case 'l': l++; break;
                case 'o': o++; break;
                case 'n': n++; break;
            }
        }

        // Since 'l' and 'o' appear twice in "balloon"
        l /= 2;
        o /= 2;

        return min({b, a, l, o, n});
    }
};

File: 1297-maximum-number-of-balloons/maximum-number-of-balloons.java
class Solution {
    public int maxNumberOfBalloons(String text) 
    {
        Set<Character> set = Set.of('b', 'a', 'l', 'o', 'n');
        Map<Character, Integer> map = new HashMap<>();

        for(char ch : text.toCharArray())
        {
            if(set.contains(ch))
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }

        // text = "lloo", "balloo", "allnnn"
        if(map.size() < 5)
        return 0;

        int instances = Integer.MAX_VALUE;
        for (Map.Entry<Character, Integer> entry : map.entrySet()) 
        {
            char key = entry.getKey();
            int value = (key == 'l' || key == 'o') ? entry.getValue() / 2 : entry.getValue();

            instances = Math.min(instances, value);
        }
        return instances;
    }
}